bionic/code_analyzer.py

```diff
@@ -0,0 +1,181 @@
+"""
+This module contains the CodeHasher class, which implements the logic
+to hash code objects into a value that unqiuely identifies the objects.
+
+The CodeHasher class is used by Bionic to detect any changes to entity
+functions. When an entity function changes, the value of the Bionic
+entity may also change, which requires Bionic to invalidate the cache
+and compute the entity again.
+
+This class is only used when versioning mode is set to "auto" or
+"assisted". Since the logic doesn't cover every type of code change,
+like references or classes, these versioning modes are still
+experimental. When it reaches a state where it detects all kinds of
+changes, we will make this the default behavior.
+"""
+
+import hashlib
+import inspect
+import json
+import sys
+import warnings
+
+from enum import Enum
+
+from .utils.misc import oneline
+
+
+DEFAULT_SIZE = 10
+
+
+class CodeHasher:
+    """
+    This class hashes code objects into a value that uniquely
+    identifies the objects. If a code object is different from what was
+    seen previously, analyzing it would result in a different hash.
+
+    For now, we only examine the function bytecode and associated
+    constants; in the future, we will add more functionality like
+    checking function references and any changes to class objects.
+    """
+
+    def __init__(self):
+        self._hash = hashlib.new("md5")
+        # This is used to detect circular references.
+        self._object_depth_by_id = {}
+
+    def hash(self, obj):
+        self._update_with_checks(obj=obj, depth=0)
+        return self._hash.hexdigest()
+
+    def _update_with_checks(self, obj, depth):
+        """
+        Checks for circular references before calling _update method,
+        which does the actual encoding.
+        """
+        obj_id = id(obj)
+
+        # If the obj is already being hashed, break the circular ref by
+        # analyzing the depth of the value instead.
+        if obj_id in self._object_depth_by_id:
+            self._hash.update(CodeHashPrefix.CIRCULAR_REF.value)
+            obj = self._object_depth_by_id[obj_id]
+            obj_id = id(obj)
+
+        self._object_depth_by_id[obj_id] = depth
+        self._update(obj, depth)
```

Comment:
Wouldn't it be easier to increment the depth here, rather than every place where you call `_update_with_checks`?

Reply:
Actually, do we even need to track `depth`, or is it the same as the size of `_object_depth_by_id`?

Reply:
We need to track depth since we are hashing the depth of the reference, not the current obj. What I mean is, for a chain like `a -> b -> c -> b`, depth of `b` when hashing `c` should be 1 (depth of `a` is 0). If we use the size of `_object_depth_by_id`, depth will be 2, which is the depth of `c`, not `b`. 

In short, we need to track the depth. If we don't, then `a -> b -> c -> b` and `a -> b -> c -> a` will return the same hash.

Reply:
Added a test for this.

Reply:
Sorry, "track `depth`" was ambiguous. I agree we need to track the depth of every in-progress object. However, do we need to pass around the current depth? It seems like it should always be basically equal to the number of in-progress objects.